Animation feedback on hitting correct moles, and incorrect moles

lucasmnt opened the pull request:
Addition of two new animations : 
- PopCorrectMole.anim => plays when you hit a correct mole
- PopWrongMole.anim => plays when you hit a distractor mole, or when a correct mole disappear via its timer

![image](https://user-images.githubusercontent.com/94845309/195832425-c04c0885-255c-4eb6-8f00-221e536d5561.png)


Modification of the diskMole prefab :
- Addition of a ImagePlaceholder game object
- Adding checkmark-correct and checkmark-wrong as child of ImagePlaceholder.

![image](https://user-images.githubusercontent.com/94845309/195831993-2e4ee092-8d1e-43f6-9d69-d09fe8a94e5e.png)


Modification of the diskMole.cs script to be able to play the new animations : not much has really changed in the code, beside some minor changes in the name of animations to play, depending on the mole state :

![image](https://user-images.githubusercontent.com/94845309/195833128-79c75d84-3ecc-461c-a363-8b1de93699df.png)


Here is an exemple of what PopCorrectMole.anim and PopWrongMole.anim will look like in-game : 

![MoleFeedbackBigger](https://user-images.githubusercontent.com/94845309/195832866-4640fab2-1931-4dff-962a-99ac06e92066.gif)

bastianilso commented:
Hi @lucasmnt !

- I like how the behavior looks and the general direction you ended up choosing.
- transitions and feedback duration are really well timed, good job 👍

- The "metaphor" you are in the sense relying on here is that the "mole turns into a checkmark or a cross if you hit the right or wrong one." But it looks odd that you can see some kind of remnants of mole behind it - it looks like it's 95% fully covered, but the last 5% makes it seem like you're not supposed to be able to see the mole behind it. 
![image](https://user-images.githubusercontent.com/3967945/195844675-eb20a97d-f76f-4d78-9080-69a6b7156759.png)
Actually, here is an idea: 
 1. We separate the ✔❌ icons from color. The ✔❌ are overlaid on top of the moles.
 1. We can then apply the color change on the Mole itself, following a similar timing as the icon by controlling transition logic from `Mole.cs` itself. I think that in terms of visual look, this will result in something cleaner, than having "two circles on top of each other".

- I think it works OK that the checkmark state is green when it's for such a short time. But we should aim for a lighter type of green, try with the color: 96d47e.

lucasmnt commented:
Hi @bastianilso, 

I'm working on animation feedback, and I'm trying to get it work like you suggested, playing with the main texture of the mole and the color, but the only thing I manage to do right now is this : 

![MolefeedbackV2Sad](https://user-images.githubusercontent.com/94845309/196133666-0097ec89-c154-4349-a109-47a637b41647.gif)


On the other hand, i've modified the existing animation a bit, and re-scaled the checkmark/cross circles so we cannot see what is behind : 

![MolefeedbackV2notSad](https://user-images.githubusercontent.com/94845309/196136321-cf7ce0a9-4055-47aa-8014-22366d54c566.gif)


I feel that this is not much but still a little better :)

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -80,36 +86,43 @@ public void EndPlayPop()
+            meshMaterial.mainTexture = textureEnabled;
         }
         else if (moleType == Mole.MoleType.DistractorLeft)
         {
             meshMaterial.color = fakeEnabledColor;
-            meshMaterial.mainTexture =  distractorLeftTexture;
-        } else if (moleType == Mole.MoleType.DistractorRight)
+            meshMaterial.mainTexture = distractorLeftTexture;
+        }
+        else if (moleType == Mole.MoleType.DistractorRight)
         {
             meshMaterial.color = fakeEnabledColor;
-            meshMaterial.mainTexture =  distractorRightTexture;
+            meshMaterial.mainTexture = distractorRightTexture;
```
generaly readability: do your reviewer a favor and keep changes not related to what you try to do out of the commit. In general, we should avoid making changes to tabulations, because everybody's editors are set up differently. :)

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -80,36 +86,43 @@ public void EndPlayPop()
     {
         PlaySound(enableSound);
-        PlayAnimation("EnableDisable");
+        if (moleType == Mole.MoleType.Target)
+        {
+            StartCoroutine(PlayAnim("PopWrongMole"));
+        }
+        else
+        {
+            StartCoroutine(PlayAnim("EnableDisable"));
+        }
```
Given the logic implied here, I would expect that if one animation is called "PopWrongMole", the other would be called "PopCorrectMole". But I see that you also have logic inside `PlayPop()` like this? In that case its not very clear to me why we need another if/else in the "EnableDisable" function.

lucasmnt replied:
This bit of code here allows correct moles that disables themselves through their timer to disappear with a "PopWrongMole" animation, that means that in the end of their timer, they will disappear and you'll see the cross.
I felt like having the "PopCorrectMole" animation for distractor moles that de-spawn by themselves was not pertinant.
That's why I kept "EnableDisable" animation here.

bastianilso replied:
aaah, thanks this makes a lot of sense. I think your judgement was very good, I agree that giving some kind of indication for expired moles is a good idea, given that players were supposed to hit them. I I didnt consider this corner case from a design perspective, which I guess have several questions:
 - should we use the same feedback as when people hit moles? will this cause people to interpret it as "hits" somehow? I hope not.
 - is it weird if we show feedback in the mole expired case, but not in distractor expired case?
 
Regardless, I am fine with your current judgement of the behavioral design. But since the behavior does not read very easily from a technical perspective (just by looking at the if statement), please add an inline code comment describing that this adds the PopWrongMole animation to a Mole if it expires before players manage to hit it.

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -80,36 +86,43 @@ public void EndPlayPop()
+        }
         meshMaterial.color = disabledColor;
-        meshMaterial.mainTexture =  textureDisabled;
+        meshMaterial.mainTexture = textureDisabled;
```
In general, we should avoid making changes to tabulations, because everybody's editors are set up differently. :)

lucasmnt replied:
I see, my bad, I'll make sure to not to mess with tabulation in general in the code.

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -200,9 +220,9 @@ private void SwitchShader(bool glow = false)
     // Ease function, Quart ratio.
-    private float EaseQuartOut (float k) 
+    private float EaseQuartOut(float k)
     {
-        return 1f - ((k -= 1f)*k*k*k);
+        return 1f - ((k -= 1f) * k * k * k);
     }
```
keep style changes out thx

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -230,4 +250,10 @@ private IEnumerator TransitionColor(float duration, Color startColor, Color endC
         // When the animation is finished, resets the color to its end value.
         ChangeColor(endColor);
     }
+
+    private IEnumerator PlayAnim(string animName)
+    {
+        PlayAnimation(animName);
+        yield return null;
+    }
```
Is it strictly necessary to play the animation from an IEnumerator? I see other places in DiskMole.cs where we just play the animation straight from the class itself.

bastianilso reviewed (changes requested):
@lucasmnt thanks for investigating my suggestion - do feel free to continue investigating it if you want. 

I think your other implementation could work too and the behavior looks much better now that you have ensured that it covers the mole completely. I still think the green color in the PopCorrectMole animation could be brighter.

I have taken a look at your commit code, see comments.

bastianilso commented:
Thanks for your inline replies @lucasmnt, see my new comment and let me know when the branch is ready for final review.

lucasmnt commented:
> aaah, thanks this makes a lot of sense. I think your judgement was very good, I agree that giving some kind of indication for expired moles is a good idea, given that players were supposed to hit them. I I didnt consider this corner case from a design perspective, which I guess have several questions:
> 
>     * should we use the same feedback as when people hit moles? will this cause people to interpret it as "hits" somehow? I hope not.
> 
>     * is it weird if we show feedback in the mole expired case, but not in distractor expired case?
> 
> 
> Regardless, I am fine with your current judgement of the behavioral design. But since the behavior does not read very easily from a technical perspective (just by looking at the if statement), please add an inline code comment describing that this adds the PopWrongMole animation to a Mole if it expires before players manage to hit it.

First, thank you for your feedback,

About your questions : 

- I think that if we do more differents feedbacks it could be "dangerous" in a sense that it could get confusing for the patient as too many differents feedback will lead to too much information to retain and react to (and I assume we don't want that). I also want to add that our checkmark and cross images are generic enough to have multiple use cases, so people could register the feedback as "good behavior/bad behavior". I guess we will need some field testing to have the correct awnser or at least what is the closest to it :)
- I would say this is kind of the same awnser for this question, I don't think we want too many different feedbacks for the patient, and plus, for example, using "PopCorrectMole" for exipiring distractor moles feels a bit off. From a "gamer" perspective, since you are not earning "points" by avoiding distractors, I don't feel like it is important to show that you succesfully avoided it. I consider that having "PopWrongMole" showing when you hit a distractor will be enough.

I added some commentaries so now getting inside the code will be easier.

Also i modified the color to this green : 96d47e : 

![MolesfeedbackV3color](https://user-images.githubusercontent.com/94845309/196626421-d8f0b366-4880-45e5-99df-f93fd7c0dc6e.gif)

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -78,38 +84,45 @@ public void EndPlayPop()
-        PlayAnimation("EnableDisable");
+        if (moleType == Mole.MoleType.Target) 
+        {
+            PlayAnimation("PopWrongMole"); // This animation will play if a correct mole expires via its timer
```
Rather than writing what the code does, I would like your code comment to document why the code does it.  For example
```
// Show negative feedback to users when a correct moles expires, to make it clear that they missed it
```

bastianilso commented on Assets/Scripts/Moles/DiskMole.cs:
```diff
@@ -78,38 +84,45 @@ public void EndPlayPop()
     */
 
-    protected override void PlayEnabling()
+    protected override void PlayEnabling() // plays when the wall and the moles have already spawned, when the different kind of moles are activating
```
Please keep code comments one line above the functions (this is the convention we follow in other places of the Whack-A-Mole code).

bastianilso reviewed (changes requested):
thanks for the code comment, lucas. Two more minor comments

bastianilso reviewed (changes requested):
- try discarding changes to MainScene and see if your pull request still works as intended (changes to MainScene should not be necessary, unless im mistaken in which case let me know what has changed).
- disable tabular/space changes in your editor if possible, thx!

lucasmnt commented:
I reverted changes of the main scene, and everything seems to work as intended.
I'll see what I can do about about tabular/space changes !

bastianilso commented:
@lucasmnt OK, sounds good. let me know when youre done and I will take a final look to see if we can land this.

lucasmnt commented:
Well, I have not much to add, so I think it is good to merge, without the main scene. I went to my IDE settings, and I think it will not add spaces by itself.

bastianilso commented:
OK great Lucas. Right now it looks like the branch consists of 5 commits, so maybe you can squash them into one,  make sure the MainScene is not included in there and make sure the commit is rebased on top of the master branch.

lucasmnt commented:
I add some troubles with rebasing the commits, but now it is done and normally it should work as usual

lucasmnt commented:
Closing it for the moment.
